Use env var in spark container

owenowenisme opened the pull request:
<!--
    Possible security vulnerabilities: STOP here and contact security@apache.org instead!

    Please update the title of the PR with a meaningful message - do not leave it "empty" or "generated"
    Please update this summary field:

    The summary should cover these topics, if applicable:
    * the motivation for the change
    * a description of the status quo, for example the current behavior
    * the desired behavior
    * etc

    PR checklist:
    - Do a self-review of your code before opening a pull request
    - Make sure that there's good test coverage for the changes included in this PR
    - Run tests locally before pushing a PR (./gradlew check)
    - Code should have comments where applicable. Particularly hard-to-understand
      areas deserve good in-line documentation.
    - Include changes and enhancements to the documentation (in site/content/in-dev/unreleased)
    - For Work In Progress Pull Requests, please use the Draft PR feature.

    Make sure to add the information BELOW this comment.
    Everything in this comment will NOT be added to the PR description.
-->
According to comment from @eric-maynard https://github.com/apache/polaris/pull/1471#discussion_r2065710938, it is not a good practice to use `sed` to manipulate docker-compose.yml, so I changed to use env var instead.

```bash
> export USER_CLIENT_ID="XXXX"                                                                                 
> export USER_CLIENT_SECRET="YYYY"                                                   
> docker compose -f getting-started/eclipselink/docker-compose.yml  config spark-sql 
name: eclipselink
services:
  polaris:
  # omit
  polaris-setup:
  # omit
  spark-sql:
    command:
      - /opt/spark/bin/spark-sql
      - --packages
      - org.apache.iceberg:iceberg-spark-runtime-3.5_2.12:1.7.0,software.amazon.awssdk:bundle:2.28.17,software.amazon.awssdk:url-connection-client:2.28.17,org.apache.iceberg:iceberg-gcp-bundle:1.7.0,org.apache.iceberg:iceberg-azure-bundle:1.7.0
      - --conf
      - spark.sql.extensions=org.apache.iceberg.spark.extensions.IcebergSparkSessionExtensions
      - --conf
      - spark.sql.catalog.quickstart_catalog=org.apache.iceberg.spark.SparkCatalog
      - --conf
      - spark.sql.catalog.quickstart_catalog.type=rest
      - --conf
      - spark.sql.catalog.quickstart_catalog.warehouse=quickstart_catalog
      - --conf
      - spark.sql.catalog.quickstart_catalog.uri=http://polaris:8181/api/catalog
      - --conf
      - spark.sql.catalog.quickstart_catalog.credential=XXXX:YYYY # env var substitution works
      - --conf
      - spark.sql.catalog.quickstart_catalog.scope=PRINCIPAL_ROLE:ALL
      - --conf
      - spark.sql.defaultCatalog=quickstart_catalog
      - --conf
      - spark.sql.catalogImplementation=in-memory
      - --conf
      - spark.driver.extraJavaOptions=-Divy.cache.dir=/tmp -Divy.home=/tmp
 # omit
```

MonkeyCanCode commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -172,9 +172,8 @@ Finally, note that we include the `hadoop-aws` package here. If your table is us
-USER_CLIENT_ID="XXXX"
-USER_CLIENT_SECRET="YYYY"
-sed -i "s/^\(.*spark\.sql\.catalog\.quickstart_catalog\.credential=\).*/\1${USER_CLIENT_ID}:${USER_CLIENT_SECRET}\",/" getting-started/eclipselink/docker-compose.yml
+export USER_CLIENT_ID="XXXX"
```
@eric-maynard 
Maybe use `.env` instead?

owenowenisme replied:
I thought this is a demonstration page? Isn't this a one-time thing?

MonkeyCanCode replied:
Fair enough. What I was thinking is, it will be 2 less lines to copy/paste and we put some default credential such as `root:s3cr3t` in the `.env` and add a note said to use a diff credential, update `.env` etc.

eric-maynard replied:
.env is a good idea, I wish I would have thought of that!

@owenowenisme you're right that this is a one-time thing... but that's also why modifying the `getting-started/eclipselink/docker-compose.yml` is suspect. 

Maybe we could write to a new file and `source` it instead of using the normal .env? Or we could just append to the end of the existing .env.

owenowenisme replied:
Yeah we could use `.env` file.
And I'll use `SPARK_USER_CLIENT_ID` `SPARK_USER_CLIENT_SECRET` as varName.

WDYT?

adnanhemani replied:
I'd just use `USER_CLIENT_ID` and `USER_CLIENT_ID` - and we should extend this change to Trino below as well. We're doing the same thing there.

adnanhemani replied:
If we are doing `.env` - let's also make sure to add it to the `.gitignore` or else we land into the same problem of users might accidentally commit their credentials using that file instead 😅

owenowenisme commented:
Some summary here:
- Modified `iceberg.properties` to use env var in trino
- Modified `create-catalog.sh` to use env var when creating quickstart catalog
- Modified `docker-compose.yml` pass env into container
- Add more instruction in `using-polaris.md` to demonstrate using .env

eric-maynard commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -21,7 +21,12 @@ Title: Using Polaris
 ---
-
+## Before start
+Put your `CLIENT_ID` & `CLIENT_SECRET` in .env file for later use.
```
This is literally the first line in the quickstart guide. I don't know what these things are.

owenowenisme replied:
Added more explicit instruction.

owenowenisme commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -79,11 +84,13 @@ With a catalog created, we can create a [principal]({{% relref "../entities#prin
 ./polaris ... principals create example
 {"clientId": "XXXX", "clientSecret": "YYYY"}
+echo "USER_CLIENT_ID="XXXX"                                                
+USER_CLIENT_SECRET="YYYY"" >> .env
```
Thanks, just fixed them.

adnanhemani reviewed (approved):
LGTM, thanks for this! One thing that we should fast-follow if we do not include in this PR itself is adding the instructions of the environment file to the Getting Started on EclipseLink and JDBC READMEs: 

* https://github.com/apache/polaris/blob/main/getting-started/jdbc/README.md
* https://github.com/apache/polaris/blob/main/getting-started/eclipselink/README.md

owenowenisme commented:
@adnanhemani Got it! I'll work on that as soon as this is merged.

eric-maynard commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -21,7 +21,12 @@ Title: Using Polaris
 ---
-
+## Setup
+Define your `CLIENT_ID` & `CLIENT_SECRET` and put them in .env file for later use.
```
Wait sorry, I just realized that this is already detailed in the [quickstart](https://polaris.apache.org/in-dev/unreleased/getting-started/quickstart/) page which links into here:

```
When Polaris is run using the ./gradlew run command, the root principal credentials are root and secret for the CLIENT_ID and CLIENT_SECRET, respectively. You can also set these credentials as environment variables for use with the Polaris CLI:

export CLIENT_ID=root
export CLIENT_SECRET=secret

```

So we now have redunandant but inconsistent instructions when going from the quickstart into `using-polaris`

eric-maynard replied:
Can't we just keep suggesting that users `export` rather than echo things into `.env` here?

.env is useful for when we're trying to automate things, but it seems overcomplicated to have to do all this just to set an env var:
```
echo "USER_CLIENT_ID=XXXX" >> .env
echo "USER_CLIENT_SECRET=YYYY" >> .env
cd polaris
set -a
source .env
cd spark
```

owenowenisme replied:
I don't think this is conflicting.
Using Polaris is another page.

owenowenisme replied:
> Can't we just keep suggesting that users `export` rather than echo things into `.env` here?
> 
> .env is useful for when we're trying to automate things, but it seems overcomplicated to have to do all this just to set an env var:
> 
> ```
> echo "USER_CLIENT_ID=XXXX" >> .env
> echo "USER_CLIENT_SECRET=YYYY" >> .env
> cd polaris
> set -a
> source .env
> cd spark
> ```

I originally is use export, so you prefer changing it back?

eric-maynard replied:
I think export makes sense for places where we want the user to manually manage credentials within a session so that they can be used later on. .env makes sense if we are automating something and need to manage credentials on the user's behalf.

owenowenisme commented:
@eric-maynard PTAL🙏

adnanhemani commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -236,9 +235,6 @@ org.apache.iceberg.exceptions.ForbiddenException: Forbidden: Principal 'quicksta
 Replace the credentials used in the Docker container using the following code:
```
```suggestion
Refresh the Docker container with the user's credentials:
```

adnanhemani commented on site/content/in-dev/unreleased/getting-started/using-polaris.md:
```diff
@@ -154,27 +161,19 @@ bin/spark-sql \
 
 Finally, note that we include the `hadoop-aws` package here. If your table is using a different filesystem, be sure to include the appropriate dependency.
 
 #### Using Spark SQL from a Docker container
```
```suggestion
#### Using Spark SQL from a Docker container

Refresh the Docker container with the user's credentials:
```

adnanhemani commented on getting-started/eclipselink/docker-compose.yml:
```diff
@@ -79,7 +82,7 @@ services:
       --conf, "spark.sql.catalog.quickstart_catalog.warehouse=quickstart_catalog",
       --conf, "spark.sql.catalog.quickstart_catalog.uri=http://polaris:8181/api/catalog",
-      --conf, "spark.sql.catalog.quickstart_catalog.credential=root:s3cr3t",
+      --conf, "spark.sql.catalog.quickstart_catalog.credential=${USER_CLIENT_ID}:${USER_CLIENT_SECRET}",
```
Given that the USER_CLIENT_ID won't be defined when the Docker Compose is first started, have you tested that this container will still start without issues?

owenowenisme replied:
Yes I've tested and it worked!

adnanhemani reviewed (approved):
LGTM - thanks @owenowenisme!

singhpk234 commented:
The assets are used for jdbc as well, and as result the JDBC quick start i completely broken @eric-maynard @adnanhemani.

will be putting in a quick fix shortly for that.
